### PR TITLE
Feature/side effect handling

### PR DESCRIPTION
**Pull request by mttwc**
- Implement a standard means of handling side effects
  - Components are pure. This is a discipline we really want to enforce going forward.
  - Callbacks to signal to the parent application of important changes (oneNotePickerCallbacks.ts)
  - Global props:
    - Standardize a way to get OneNote data (e.g., from API/mocking/other sources) (oneNoteDataProvider.ts)
    - Standardize a way to update the local notebook tree without losing information such as isExpanded (notebookListUpdater.ts)
    - Callbacks to the parent app (as mentioned above)
    - Selected id
  - Side effects flow in a unidirectional manner. The component that wants to call some side-effectey code does so, then calls one of the global callbacks with the result. The changes then propogate throughout the component hierarchy the react-way.
- In theory a lot of these classes were made to be unit testable or mockable/extendable. Once the test runner stuff is up I'll add UTs.
- Created some data structures for stuff like notebook, section page. Why not reuse OneNoteApi? I don't think we need all the metadata that comes in the response, and by stripping it down before piping it into the component tree, it makes debugging and testing (mocking) way simpler.
- Updated sample app. When the user clicks on a section, we now fetch a mocked out request to get a sample page, then update the local notebook tree and re-render using the new callback/side-effect logic.

### CODE REVIEW

**Comment by leakymemory** [comment]
>Why not reuse OneNoteApi? I don't think we need all the metadata that comes in the response, and by stripping it down before piping it into the component tree, it makes debugging and testing (mocking) way simpler.

I agree.  This is probably a good opportunity to re-visit the OneNoteApi and see if we can't clean things up a little.

**Review by leakymemory — APPROVED** [review]

**Comment by mttwc** [comment]
Merged in the test changes by @anbinm-msft. Will write some tests before merging this stuff in.

**Comment by mttwc** [comment]
Merging this to unblock people who might need this branch's changes. Will continue to write UTs in another branch.

**Pull request merged**
